### PR TITLE
Enhancements for upgrades

### PR DESCRIPTION
**Pull request by danielui**
1. DeviceVersions and RepositoryAPI instances are now automatically created in SoftwareActionAPI and PartitionManagerAPI constructor.
2. In case of upgrade and activate software, user can pass image path or just version.
3. Set current partition as default is possible now.
4. Add usage example for PartitionManager and SoftwareActionAPI.

### CODE REVIEW

File: vmngclient/api/software_action_api.py [hunk 1]
```diff
@@ -58,29 +60,58 @@ class SoftwareActionAPI:
...
+    session = create_vManageSession(...)
+
+    # Prepare devices list
+    devices = [dev for dev in DevicesAPI(session).devices if dev.hostname in ["vm5", "vm6"]]
+    software_image = PurePath("c8000v-universalk9.17.06.03a.0.56.SPA.bin")
+
+    # Upgrade
+    software_action = SoftwareActionAPI(session,DeviceCategory.VEDGES.value)
+    software_action_id = software_action.upgrade_software(devices,
+        InstallSpecHelper.CEDGE.value, reboot = False, sync = True, software_image=software_image)
```
Comment (igosoft) [inline vmngclient/api/software_action_api.py:75]: Can we get rid of InstallSpecHelper.CEDGE.value and replace by DeviceCategory ?

**Review by sbasan — CHANGES_REQUESTED** [review]
Provide better type hints if possible

File: vmngclient/api/software_action_api.py [hunk 2]
```diff
@@ -58,29 +60,58 @@ class SoftwareActionAPI:
...
+    Usage example:
+    # Create session
+    session = create_vManageSession(...)
+
+    # Prepare devices list
+    devices = [dev for dev in DevicesAPI(session).devices if dev.hostname in ["vm5", "vm6"]]
+    software_image = PurePath("c8000v-universalk9.17.06.03a.0.56.SPA.bin")
+
+    # Upgrade
+    software_action = SoftwareActionAPI(session,DeviceCategory.VEDGES.value)
```
[the remark this answers is not in the document]
Comment (danielui) [reply vmngclient/api/software_action_api.py:73]: endpoint for that value is: "/dataservice/system/device/vedges". I think, it's good idea to name enum, the same like value. If I'm wrong, I can change it.

File: vmngclient/api/partition_manager_api.py [hunk 3]
```diff
@@ -70,7 +107,7 @@ def remove_partition(self, devices: List[Device], version: str, force: bool = Fa
     def _check_remove_partition_possibility(self, devices) -> List:
```
[the remark this answers is not in the document]
Comment (danielui) [reply vmngclient/api/partition_manager_api.py]: Done.

File: vmngclient/api/software_action_api.py [hunk 4]
```diff
@@ -182,9 +224,7 @@ def _downgrade_check(self, devices, version_to_upgrade: str, family) -> List:
             Union[None, List]: [None, list of devices with no permission to downgrade]
```
[the remark this answers is not in the document]
Comment (danielui) [reply vmngclient/api/software_action_api.py]: Done.

File: vmngclient/api/versions_utils.py [hunk 5]
```diff
@@ -144,12 +145,30 @@ def get_device_list_in_available(self, version_to_set_up: str, devices: List[Dev
...
         Args:
             version_to_set_up (str): requested version
+            devices (List[Device]): devices on which action going to be performed
 
         Returns:
             list : list of devices
         """
         return self._get_device_list_in(version_to_set_up, devices, "available_versions")
 
+    def get_devices_current_version(self, devices: List[Device]) -> List[dict]:
```
Comment (sbasan) [inline vmngclient/api/versions_utils.py:155]: Looks like all methods in this module uses List[dict] but data is structured
Is it possible to use typed list?
eg. List[DeviceId]
where new dataclass is declared in this module:
```python
@dataclass
class DeviceId():
  deviceId: str
  deviceIP: str
```

**Review by sbasan — CHANGES_REQUESTED** [review]
improve type hinting #2

File: vmngclient/api/versions_utils.py (hunk 5 above)
Comment (danielui) [reply vmngclient/api/versions_utils.py:155]: It's possible, but it's scope for the next PR.

**Review by sbasan — APPROVED** [review]

**Pull request merged**